TASK-15 Candidate testing phase

uncmmath opened the pull request:
+ Add testing phase states(switching from core service): 
  - PENDING: accepting candidate applications, testing phase not started yet 
  - ACTIVE: candidate application phase closed, testing phase started. When this phase starts, TestCreatedEvent messages are sent via kafka in order to notify candidates using mailService When candidate passes the test, TestPassedEvent messages are sent via kafka 
  - COMPLETE: testing phase completed, test passing is no longer allowed

- Some dto refactoring and other minor improvements

uncmmath committed:
TASK-15 Candidate testing phase

- Add testing phase states(switching from core service):
    - PENDING: accepting candidate applications, testing phase not started yet
    - ACTIVE: candidate application phase closed, testing phase started.
       When this phase starts, TestCreatedEvent messages are sent via kafka in order to notify candidates using mailService
       When candidate passes the test, TestPassedEvent messages are sent via kafka
    - COMPLETE: testing phase completed, test passing is no longer allowed

- Some dto refactoring and other minor improvements

Ivanesca commented on mail/build.gradle.kts:
```diff
@@ -22,8 +22,8 @@ dependencies {
     implementation("org.springframework.kafka:spring-kafka")
-    implementation(project(":domain"))
     testImplementation("org.springframework.boot:spring-boot-starter-test")
+    implementation(project(":domain"))
```
Точно ли нужно тянуть весь domain в  mail?

uncmmath replied:
Ну, в качестве альтернативы можно создать ещё один модуль чисто для классов, реализующих IMailEvent, но мне кажется, что это уже прямо микрооптимизации

uncmmath replied:
В противном случае просто придется дублировать классы ивентов, что выглядит не очень удобным

uncmmath committed:
TASK-15 Candidate testing phase

- Retry reading from kafka topic on exceptions (e.g. database errors)
- Improve testVariant finishing state: add TestVariantFinalizerService that emits TestVariantStateFinished event on answering last question of test or on timeout (when test time is up)
- Send TestCreatedEvent to kafka topic when TestVariant is created in ACTIVE testing phase state

- Refactoring kafka configuration
